content/5_Plat/RURQ.mdx

```diff
@@ -152,135 +152,124 @@ Given the mechanics behind the `pushdown` function, all you need is a regular
 range-sum segment tree to solve the question.
 
 ```cpp
-
-#include <iostream>
-using ll = long long;
+#include <bits/stdc++.h>
 using namespace std;
+using ll = long long;
 
-const int maxN = 2e5 + 5;
-
-int N, Q;
-int a[maxN];
-
-struct node {
-	ll val;
-	ll lzAdd;
-	ll lzSet;
-	node(){};
-} tree[maxN << 2];
-
-#define lc p << 1
-#define rc (p << 1) + 1
+// T = tree node type, which will be long long
+// U = update type, which will be array<ll, 2>
+// Updates are: {lazy_set, lazy_add}
```

Comment:
turn into doc comment

Reply:
ok so upon further inspection it seems like you're trying to do two things at once with `U` which is... kinda sus
why not just have separate functions for setting and adding
yk, single responsibility and whatnot

Reply:
reason i do two things with `U` at once are:
- shortens the code a lot (because `range_add` and `range_set` would have 90% of the same code)
- it makes templating significantly easier (altho ig its not super applicable here)
idk, I can change it if you want but I think the way it currently is works fine

Reply:
i don't see any purpose in making it a tempalte argument when it really only takes a two-element indexable thing...maybe a custom struct at that point?

Reply:
> i don't see any purpose in making it a tempalte argument when it really only takes a two-element indexable thing...maybe a custom struct at that point?

fair enough, I can do that later. do you still want range updates to have setting and adding to be different methods?

Reply:
i mean say i set both values of the array to some nonzero number
what would happen?

Reply:
range set always takes precedence

the way i wrote the code is so that `lz[v]` would only ever have either `set` or `add` be nonzero though

Reply:
could you do something like this then

```cpp
enum QueryType {
    SET, ADD
};

struct Query {
    QueryType type;
    int val;
};
```

Reply:
ok it took me a while b/c i was being a bit of a clown, and I did have to change up some stuff, but I think it looks pretty nice now